File: Algorithms/Greedy/FractionalKnapsack.js
"use strict";

class Item {

    constructor(p, w) {

        this.price = p;
        this.weight = w;

    }

    get ratio() {

        return this.price / this.weight;

    }

}

let getMaxValue = (items, maxWeight) => {

    items.sort((a, b) => b.ratio - a.ratio);

    let currentWeight = 0,
        currentValue = 0;

    for (let item of items) {

        if (currentWeight + item.weight <= maxWeight) {

            currentWeight += item.weight;
            currentValue += item.price;

        } else {

            let remaining = maxWeight - currentWeight;

            currentValue += remaining * item.ratio;

            break;

        }

    }

    return currentValue;

}

let items = [
    new Item(60, 10),
    new Item(100, 20),
    new Item(20, 50),
    new Item(120, 30),
    new Item(10, 2)
];

console.log(getMaxValue(items, 75));

File: Algorithms/Greedy/FractionalKnapsack.py

class Item(object):

    def __init__(self, v, w):
        self.value = v
        self.weight = w

    @property
    def ratio(self):
        return self.value / float(self.weight)

def get_max_value(items, max_weight):

    items.sort(key=lambda item: item.ratio, reverse=True)

    current_weight, current_value = 0, 0.0

    for item in items:

        if current_weight + item.weight <= max_weight:

            current_weight += item.weight
            current_value += item.value

        else:

            remaining = max_weight - current_weight

            current_value += remaining * item.ratio

            break

    return current_value

def main():

    items = [
        Item(60, 10),
        Item(100, 20),
        Item(20, 50),
        Item(120, 30),
        Item(10, 2)
    ]

    print(get_max_value(items, 75))

if __name__ == '__main__':
    main()
